src/main/cpp/src/KudoGpuSerializerJni.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cudf_jni_apis.hpp"
+#include "shuffle_split.hpp"
+
+extern "C" {
+
+JNIEXPORT jlongArray JNICALL
+Java_com_nvidia_spark_rapids_jni_kudo_KudoGpuSerializer_splitAndSerializeToDevice(
+  JNIEnv* env, jclass, jlong j_table_view, jintArray j_splits)
+{
+  JNI_NULL_CHECK(env, j_table_view, "table is null", NULL);
+  JNI_NULL_CHECK(env, j_splits, "splits is null", NULL);
+  try {
+    cudf::jni::auto_set_device(env);
+
+    auto table = reinterpret_cast<cudf::table_view const*>(j_table_view);
+    const cudf::jni::native_jintArray n_splits(env, j_splits);
+    std::vector<cudf::size_type> splits = n_splits.to_vector<int>();
+
+    auto [split_result, split_meta] = spark_rapids_jni::shuffle_split(
+      *table, splits, cudf::get_default_stream(), cudf::get_current_device_resource());
+    // We need to copy the offsets back to a host memory buffer
+
+    cudf::jni::native_jlongArray result(env, 6);
+    result[0]    = reinterpret_cast<jlong>(split_result.partitions->data());
+    result[1]    = static_cast<jlong>(split_result.partitions->size());
+    result[2]    = reinterpret_cast<jlong>(split_result.partitions.get());
+    auto offsets = std::make_unique<rmm::device_buffer>(std::move(split_result.offsets.release()));
+    result[3]    = reinterpret_cast<jlong>(offsets->data());
+    result[4]    = static_cast<jlong>(offsets->size());
+    result[5]    = reinterpret_cast<jlong>(offsets.get());
+
+    split_result.partitions.release();
+    offsets.release();
```

Comment:
Am I reading this correctly that you're returning a pointer into offsets, but then releasing offsets here? Shouldn't the release wait until after the data has been copied out of that pointer?

Reply:
What was I thinking???